rclcpp/include/rclcpp/node_impl.hpp

```diff
@@ -92,13 +95,40 @@ Node::create_subscription(
   const SubscriptionOptionsWithAllocator<AllocatorT> & options,
   typename MessageMemoryStrategyT::SharedPtr msg_mem_strat)
 {
-  return rclcpp::create_subscription<MessageT>(
-    *this,
-    extend_name_with_sub_namespace(topic_name, this->get_sub_namespace()),
-    qos,
-    std::forward<CallbackT>(callback),
-    options,
-    msg_mem_strat);
+  if (options.topic_stats_options.state == rclcpp::TopicStatisticsState::Enable) {
+    std::shared_ptr<Publisher<statistics_msgs::msg::MetricsMessage>> publisher =
+      this->create_publisher<statistics_msgs::msg::MetricsMessage>(
+      options.topic_stats_options.publish_topic,
+      qos);
+
+    auto sub_topic_stats = std::make_shared<
+      rclcpp::topic_statistics::SubscriptionTopicStatistics<CallbackMessageT>
+      >(this->get_name(), publisher);
+
+    auto timer = this->create_wall_timer(
```

Comment:
Is it really the best idea to have a timer and publisher for each subscription with this setting turned on?

The most common use case in ROS 1 was to turn on statistics for all the topics in the graph so you could introspect multiple things simultaneously and see which topics were using the most resources or lagging behind (you didn't know ahead of time which you wanted to look at), in which case that is a lot of extra timers and publishers... Really these should share a resource for this, no?

Each subscription could collect statistical info asynchronously (as callbacks are called) and a single timer could trigger to collect them all and publish them all on a single publisher. I think that would be a much better approach, would it not?

Reply:
My fear here was the complexity of the approach and use case. 

Per the global timer: I didn't think this made since given the variety of potential topic publishing periods. For example, the default 1 second publishing period makes sense for something collected at a high frequency, but may not for much lower frequencies. 

We could have a global publisher and this was considered, but for the immediate approach I chose a per subscription publisher for ease of lifetime management. 

I think we could do the following: have a global publisher for topic statistics, where the publisher is created and held by the node, but have separate timers. How does that sound?

Reply:
Sure ok, that's definitely better.

I see now what you mean about the timers, since each subscription can set its statistics publish rate separately. I wonder though if that make sense now (connecting the publish rate with the subscription). Why not have one publisher and timer per node and have it publish all the statistics at once? What use case(s) supports the design decision to publish them separately and asynchronously? Do you know off hand how it was done in ROS 1? Was there one publisher that gathered all the stats and published them in bulk periodically or a separate publishing event for each contributor of stats (each subscription)?

An in between would be to have the Subscriptions register themselves with a "statistics publisher" class, which periodically queries all registered subscriptions and publishes their stats together, then you could have as many or few of those statistic publishers as you want, and have a default one in the node which is used if nothing else is specified. This is kind of how callback groups work, you can create one and pass it when making subscriptions, but if you don't specify one the default one in the node is used. This could also be extended more easily to incorporate more sources of stats, things other than subscriptions, in the future. They would just register themselves with a statistics publisher, collect stats asynchronously, and wait to be queried by the statistics publisher.

I think by far the dominant use case will be to specify nothing when creating the subscription, and then want to turn on stats externally without changing any code. So I expect most users would be fine with posting stats in bulk at a moderate frequency (~1hz), even if the subscriptions receive data at a slower rate. I think it's nice to be able to further configure this (which subscriptions publish stats, how and how often they publish them), but I expect that will be an uncommon use case.

Reply:
>I see now what you mean about the timers, since each subscription can set its statistics publish rate separately. I wonder though if that make sense now (connecting the publish rate with the subscription). Why not have one publisher and timer per node and have it publish all the statistics at once? What use case(s) supports the design decision to publish them separately and asynchronously? Do you know off hand how it was done in ROS 1? Was there one publisher that gathered all the stats and published them in bulk periodically or a separate publishing event for each contributor of stats (each subscription)?

ROS1: each [StatisticsLogger](https://github.com/ros/ros_comm/blob/12eaf946597739e833b2300ef28bb079f248a9e1/clients/roscpp/include/ros/statistics.h#L89) has a [publisher](https://github.com/ros/ros_comm/blob/12eaf946597739e833b2300ef28bb079f248a9e1/clients/roscpp/src/libros/statistics.cpp#L234), but publishes the data by checking the time using its [callback](https://github.com/ros/ros_comm/blob/12eaf946597739e833b2300ef28bb079f248a9e1/clients/roscpp/src/libros/statistics.cpp#L237) instead of a separate timer.  

A single timer does make sense, but now we need to rethink ownership. The node creates the publisher and the timer on behalf of the topic statistics instance, but now the node needs to own both the publisher and timer. This places more in node_impl (unless you have thoughts otherwise) which based on your previous comments seems undesirable. 

>An in between would be to have the Subscriptions register themselves with a "statistics publisher" class, which periodically queries all registered subscriptions and publishes their stats together, then you could have as many or few of those statistic publishers as you want, and have a default one in the node which is used if nothing else is specified. This is kind of how callback groups work, you can create one and pass it when making subscriptions, but if you don't specify one the default one in the node is used. This could also be extended more easily to incorporate more sources of stats, things other than subscriptions, in the future. They would just register themselves with a statistics publisher, collect stats asynchronously, and wait to be queried by the statistics publisher.

That makes sense and would be useful for publishers, as you noted. However, could we approach this as a separate improvement?

Reply:
> This places more in node_impl (unless you have thoughts otherwise) which based on your previous comments seems undesirable.

I could be owned by `NodeBase` or `NodeTopics`, right? No need for it to be in `Node`.

> However, could we approach this as a separate improvement?

Yes, but it would definitely break API in my opinion, so I assume it will have to be G-turtle and cannot be added as a non-breaking change.

Reply:
> ROS1: each StatisticsLogger has a publisher,

One thing to note is that in ROS 1 those publishers were actually a singleton...

https://github.com/ros/ros_comm/blob/12eaf946597739e833b2300ef28bb079f248a9e1/clients/roscpp/src/libros/topic_manager.cpp#L324-L356